src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -45,15 +46,53 @@ @implementation MTRDeviceConnectivityMonitor {
 }
 
 static os_unfair_lock sConnectivityMonitorLock = OS_UNFAIR_LOCK_INIT;
+
+// Tracks number of currently active monitors. When this drops to zero, triggers timer
+// to cleanup sSharedResolverConnection after kSharedConnectionLingerIntervalSeconds.
 static NSUInteger sConnectivityMonitorCount;
 static DNSServiceRef sSharedResolverConnection;
 static dispatch_queue_t sSharedResolverQueue;
 
+// Map table solution for DNS-SD callback safety:
+// Instead of passing raw object pointers to DNS-SD callbacks (which creates use-after-free
+// vulnerabilities), we assign each MTRDeviceConnectivityMonitor a unique numerical ID and
+// pass that ID as the context. The callback then looks up the monitor object in this weak
+// map table. If the object has been deallocated, the weak reference becomes nil and the
+// callback is safely ignored. This completely eliminates race conditions between object
+// deallocation and asynchronous DNS-SD callbacks.
+static NSMapTable<NSNumber *, MTRDeviceConnectivityMonitor *> *sMonitorMap;
+static NSUInteger sNextMonitorID = 1;
```

Comment:
Probably safe to assume we will not get reuse quickly enough that callbacks will be delivered to the wrong object (the new one reusing an id)?

But given that we are casting this to void*, shouldn't this be `uintptr_t`?  Or we need static_asserts about this not being bigger than void*?  Or does NSUInteger precisely guarantee being same size as void*?

Reply:
It's probably safe about reuse since it'll have to wrap around the entire uinteger space to reuse.

The point about `uintptr_t` is valid. Let me change it.